File: #1751.maximum-number-of-events-that-can-be-attended-ii.cpp
class Solution {
public:
    int maxValue(vector<vector<int>>& events, int k) {

        std::sort(events.begin(), events.end());
        int size = events.size();
        vector<vector<int>> dp(size + 1, vector<int> (k + 1, -1));
        return solve(events, k, 0, dp, size);
    }

private:
    int solve(vector<vector<int>> &events, int k,
        int idx, vector<vector<int>>& dp, int size) {
        
        if (idx == size || k == 0) {
            return 0;
        }

        if (dp[idx][k] != -1) {
            return dp[idx][k];
        }

        int result = solve(events, k, idx + 1, dp, size);
        int curr_end = events[idx][1];
        int l = idx + 1;
        int r = size - 1;
        int idx2 = size;
        while(l <= r) {
            int mid = (l + r) / 2;
            if (events[mid][0] > curr_end) {
                idx2 = mid;
                r = mid - 1;
            } else {
                l = mid + 1;
            }
        }
        return dp[idx][k] = std::max(result,
            events[idx][2] + solve(events, k-1, idx2, dp, size));
    }
};

File: #1751.maximum-number-of-events-that-can-be-attended-ii.ts
function maxValue(events: number[][], k: number): number {
    events.sort((a,b) => a[0] - b[0]);
    const memo = new Array(events.length).fill(0).map(_ => {
            let d = new Array(k);
            d[0] = 0;
            return d;
    });
 
    const binarySearch = (start:number, startTime:number) => {
        if (start >= events.length) {
            return start;
        }

        let end = events.length;
        while (start < end) {
            const mid = start + ~~((end - start) / 2);
            if (events[mid][0] >= startTime) {
                end = mid;
            } else {
                start = mid + 1;
            }
        }
        return start;
    }

    const helper = (index:number, k_left:number):number => {
        if (index >= events.length) {
            return 0;
        }

        if (memo[index][k_left] != undefined) {
            return memo[index][k_left];
        }

        let pick = events[index][2];
        const nextIndex = binarySearch(index + 1, events[index][1] + 1);
        pick += helper(nextIndex, k_left - 1);

        let dont_pick = helper(index + 1, k_left);
        memo[index][k_left] = Math.max(pick, dont_pick);
        return memo[index][k_left];
    } 
    return helper(0, k);
};
